### DISCUSSION TITLE
Better author experience for internal linking needed.

### DISCUSSION THREAD

**Discussion by kbergha** [craftcms/ckeditor#298] [Ideas] [answered]
Linking to internal Craft entities and is in need of improvements to the authoring experience / user experience.

Creating a new link is fine, but editing an existing internal link is not good for content authors.
I am wondering if the AX / UX here really is as intended, or if I've encountered an oversight / not prioritized task to fix.

After being in contact with support about the problem, we agreed that the best way forward was to create a discussion to highlight the problem, get feedback from the community, and explain my thoughts on how to improve the situation.

My assumption is that the same is an issue for internal links to assets as well, but I haven't tested this. Edit: I have now tested this, see next comment.

### Versions
Craft Pro 5.4.6
CKEditor 4.2.0

### New internal link (this is fine)
Inserting a new internal link is fine. The AX / UX here makes sense.

#### Selecting and creating link
![internal-reference-creating-link](https://github.com/user-attachments/assets/8036837c-809a-479d-8545-e22c5108a06b)

#### Finding entry to link to
![internal-reference-finding-entry](https://github.com/user-attachments/assets/dee3f643-e486-44e2-b45c-a845510aefbf)

#### Selecting entry to link to
![internal-reference-selecting-entry](https://github.com/user-attachments/assets/d981b254-bbd7-4945-86eb-c9dc27c70548)

### Updating link (this is in need of improvements)

Now, here comes the problem. Editing an existing internal link is not good.

#### Clicking on edit on an internal link
![internal-reference-before-clicking-edit](https://github.com/user-attachments/assets/c055a8d5-1b89-488f-8943-8bcfb7fdbea6)

#### After clicking edit
The author is now presented with a (long) URL.
![internal-reference-after-clicking-edit-1](https://github.com/user-attachments/assets/c4ce103f-7e73-41d2-84b6-6ed42d837b39)

To change the link, the author now has to deal with the internal reference at the end of the URL, in this example `#entry:1@1`.
There is no reasonable way for the content author to know what the current link actually links to, and you have to take a guess, maybe based on the current link text.

![internal-reference-after-clicking-edit-2](https://github.com/user-attachments/assets/71e323d4-cf4f-4278-90e8-dbe3a48b9a42)

### My suggestion
When clicking on the edit link for an internal link, the author should be presented with the same dialogue window and state as in "Selecting entry to link to" when inserting a new link. The entry is pre-selected, and it's easy to find a new entry to link to with the normal user interface.

![internal-reference-finding-entry](https://github.com/user-attachments/assets/dee3f643-e486-44e2-b45c-a845510aefbf)

### Workaround'ish

The suggestion I got from support was to remove the link and create a new one.
Yes, this works, but you loose the ability to know what you actually linked to before.
If the link text is explicit, you might be able to take a good guess, but it's still a guess.

### Your suggestion

Any thoughts? Can it be done in an even better way?

### Related issues / discussions

This one is slightly related: https://github.com/craftcms/ckeditor/discussions/92

**Comment by kbergha** [comment]
I've tested the AX / UX for assets as well.

As with entries, selecting the asset is fine, but editing is perhaps even more challenging.

### Selecting (this is fine)

![assets-1](https://github.com/user-attachments/assets/28ff70ff-3871-4bd8-862d-0006c9532532)

### Editing (this is very challenging, in need of improvements)

![assets-2](https://github.com/user-attachments/assets/e9ac093a-44b4-45a1-aaea-aba630e852f0)

The author now has to deal with the filename (which is not too hard perhaps).
But the volume handle ("/files/" in this example)? That is not ok...
And this is in addition to the internal reference `#asset:505@1`.

They probably have no relation to the handle at all, since this is not something they see in the UI, and the volume handle may of course be very different form the name they actually see.

![assets-3](https://github.com/user-attachments/assets/c7de9086-b141-4c0c-8b21-d883d80c0702)
![assets-4](https://github.com/user-attachments/assets/94ef399a-4f72-4f6d-b262-59e47158bd6e)

### My suggestion

It's the same as for entries, you should get back the same dialogue window and state as in "Selecting (this is fine)" when editing the link.

![assets-1](https://github.com/user-attachments/assets/28ff70ff-3871-4bd8-862d-0006c9532532)

**Reply by kbergha** [reply]
@i-just - If I did not misunderstand what was presented during DotAll, the issues I've raised here will be fixed by CKEditor plugin version 5 ?

**Reply by i-just** [reply]
Hi @kbergha,

Thanks for getting in touch! Yes, the links UX is changing in v5 of our CKEditor plugin. You can see the overview of changes in [this PR](https://github.com/craftcms/ckeditor/pull/375). There’s also [this one](https://github.com/craftcms/ckeditor/pull/438) with further improvements.

As mentioned by support, while on v3 or v4 of the CKEditor plugin, the easiest way to change a link would be to remove and re-add.

**Reply by kbergha** [reply]
Awesome. Thanks.

**Comment by brandonkelly** [comment, marked as answer]
We released v5 of our CKEditor plugin today, with a new Link modal that reflects Craft’s Link fields, with the ability to also include advanced fields like “Title text”. (#375)
